File: src/main/kotlin/wemi/test/forked/TestLauncher.java
package wemi.test.forked;

import org.junit.platform.engine.DiscoverySelector;
import org.junit.platform.engine.Filter;
import org.junit.platform.engine.discovery.ClassNameFilter;
import org.junit.platform.engine.discovery.DiscoverySelectors;
import org.junit.platform.engine.discovery.PackageNameFilter;
import org.junit.platform.launcher.Launcher;
import org.junit.platform.launcher.LauncherDiscoveryRequest;
import org.junit.platform.launcher.TagFilter;
import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder;
import org.junit.platform.launcher.core.LauncherFactory;
import wemi.test.TestParameters;
import wemi.test.TestReport;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.PrintStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.LogRecord;
import java.util.logging.Logger;

/**
 * Launched by the test task in a forked process.
 * Forked process has a classpath of test classes combined with Wemi jar and JUnit launcher.
 *
 * Takes no arguments.
 *
 * Reads WHOLE stdin as [TestParameters] json.
 * Program stdout is redirected to stderr, which is also where errors are printed.
 * Whole [TestReport] is printed into stdout.
 */
public final class TestLauncher {

	/**
	 * Arbitrary non-typical byte that is emitted by the test harness, before outputting the TestReport bytes.
	 * This is done because debug tools might inject messages that would break the output.
	 */
	public static final byte MAGIC_MESSAGE_START = 14; // ASCII Shift Out
	public static final byte MAGIC_MESSAGE_END = 15; // ASCII Shift In

	/** The magic message start or end must repeat this many times to count as the delimiter. */
	public static final int MAGIC_MESSAGE_DELIMITER_REPEAT = 10;

	public static void main(String[] args) {
		int exitCode;

		try {
			final PrintStream out = System.out;
			System.setOut(System.err);

			final TestParameters testParameters = new TestParameters();
			try (DataInputStream in = new DataInputStream(System.in)) {
				testParameters.readFrom(in);
			}
			setupLogging(testParameters.javaLoggingLevel);

			final Launcher launcher = LauncherFactory.create();
			ReportBuildingListener reportBuilder = new ReportBuildingListener(testParameters.filterStackTraces);
			launcher.registerTestExecutionListeners(reportBuilder);

			final LauncherDiscoveryRequest discoveryRequest;
			{
				final LauncherDiscoveryRequestBuilder requestBuilder = new LauncherDiscoveryRequestBuilder();
				final ArrayList<DiscoverySelector> selectors = new ArrayList<>();
				for (String aPackage : testParameters.selectPackages) {
					selectors.add(DiscoverySelectors.selectPackage(aPackage));
				}
				for (String aClass : testParameters.selectClasses) {
					selectors.add(DiscoverySelectors.selectClass(aClass));
				}
				for (String method : testParameters.selectMethods) {
					selectors.add(DiscoverySelectors.selectMethod(method));
				}
				for (String resource : testParameters.selectResources) {
					selectors.add(DiscoverySelectors.selectClasspathResource(resource));
				}

				final HashSet<Path> classpathRoots = new HashSet<>();
				for (String root : testParameters.classpathRoots) {
					classpathRoots.add(Paths.get(root));
				}
				selectors.addAll(DiscoverySelectors.selectClasspathRoots(classpathRoots));
				requestBuilder.selectors(selectors);

				final ArrayList<Filter<?>> filters = new ArrayList<>();
				final List<String> classIncluded = testParameters.filterClassNamePatterns.included;
				if (!classIncluded.isEmpty()) {
					filters.add(ClassNameFilter.includeClassNamePatterns(classIncluded.toArray(new String[0])));
				}
				final List<String> classExcluded = testParameters.filterClassNamePatterns.excluded;
				if (!classExcluded.isEmpty()) {
					filters.add(ClassNameFilter.excludeClassNamePatterns(classExcluded.toArray(new String[0])));
				}

				final List<String> packagesIncluded = testParameters.filterPackages.included;
				if (!packagesIncluded.isEmpty()) {
					filters.add(PackageNameFilter.includePackageNames(packagesIncluded));
				}
				final List<String> packagesExcluded = testParameters.filterPackages.excluded;
				if (!packagesExcluded.isEmpty()) {
					filters.add(PackageNameFilter.excludePackageNames(packagesExcluded));
				}

				final List<String> tagsIncluded = testParameters.filterTags.included;
				if (!tagsIncluded.isEmpty()) {
					filters.add(TagFilter.includeTags(tagsIncluded));
				}
				final List<String> tagsExcluded = testParameters.filterTags.excluded;
				if (!tagsExcluded.isEmpty()) {
					filters.add(TagFilter.excludeTags(tagsExcluded));
				}
				requestBuilder.filters(filters.toArray(new Filter<?>[0]));

				discoveryRequest = requestBuilder.build();
			}

			launcher.execute(discoveryRequest);

			// Output
			if (!reportBuilder.isComplete()) {
				System.err.println("Test report is not complete!");
			}
			final TestReport report = reportBuilder.testReport();

			try (DataOutputStream dos = new DataOutputStream(out)) {
				for (int i = 0; i < MAGIC_MESSAGE_DELIMITER_REPEAT; i++) {
					dos.write(MAGIC_MESSAGE_START);
				}
				report.writeTo(dos);
				for (int i = 0; i < MAGIC_MESSAGE_DELIMITER_REPEAT; i++) {
					dos.write(MAGIC_MESSAGE_END);
				}
			}

			exitCode = 0;
		} catch (Throwable e) {
			System.err.println("Exception while running tests");
			e.printStackTrace(System.err);

			exitCode = 1;
		} finally {
			System.out.flush();
			System.err.flush();
		}

		System.exit(exitCode);
	}

	private static void setupLogging(Level level) {
		try {
			final Logger rootLogger = Logger.getLogger("");
			for (Handler handler : rootLogger.getHandlers()) {
				rootLogger.removeHandler(handler);
			}
			rootLogger.addHandler(new Handler() {
				@Override
				public void publish(LogRecord record) {
					final Object[] parameters = record.getParameters();
					final Throwable thrown = record.getThrown();
					final StringWriter sw = new StringWriter();
					sw.append('[').append(record.getLoggerName()).append(' ').append(record.getLevel().getName()).append("] ").append(record.getMessage());
					if (parameters != null) {
						for (Object parameter : parameters) {
							sw.append(" | ");
							try {
								sw.append(parameter == null ? "null" : parameter.toString());
							} catch (Exception ignored) {}
						}
					}
					if (thrown != null) {
						sw.append('\n');
						thrown.printStackTrace(new PrintWriter(sw));
					}
					System.err.print(sw.getBuffer());
				}

				@Override
				public void flush() {}

				@Override
				public void close() throws SecurityException {
				}
			});

			rootLogger.setLevel(level);
		} catch (Exception ignored) {}
	}
}

File: src/main/kotlin/wemi/test/forked/TestLauncher.kt
package wemi.test.forked

import com.darkyen.tproll.util.StringBuilderWriter
import org.junit.platform.engine.DiscoverySelector
import org.junit.platform.engine.Filter
import org.junit.platform.engine.TestExecutionResult
import org.junit.platform.engine.TestExecutionResult.Status.*
import org.junit.platform.engine.TestSource
import org.junit.platform.engine.discovery.ClassNameFilter.excludeClassNamePatterns
import org.junit.platform.engine.discovery.ClassNameFilter.includeClassNamePatterns
import org.junit.platform.engine.discovery.DiscoverySelectors.*
import org.junit.platform.engine.discovery.PackageNameFilter.excludePackageNames
import org.junit.platform.engine.discovery.PackageNameFilter.includePackageNames
import org.junit.platform.engine.reporting.ReportEntry
import org.junit.platform.engine.support.descriptor.ClassSource
import org.junit.platform.engine.support.descriptor.MethodSource
import org.junit.platform.launcher.EngineFilter.excludeEngines
import org.junit.platform.launcher.EngineFilter.includeEngines
import org.junit.platform.launcher.TagFilter.excludeTags
import org.junit.platform.launcher.TagFilter.includeTags
import org.junit.platform.launcher.TestExecutionListener
import org.junit.platform.launcher.TestIdentifier
import org.junit.platform.launcher.TestPlan
import org.junit.platform.launcher.core.LauncherDiscoveryRequestBuilder
import org.junit.platform.launcher.core.LauncherFactory
import wemi.test.*
import wemi.util.appendWithStackTrace
import wemi.util.readJson
import wemi.util.writeJson
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.nio.file.Path
import java.nio.file.Paths
import java.time.ZoneId

/**
 * Launched by test task in a forked process.
 * Forked process has a classpath of test classes combined with Wemi jar and JUnit launcher.
 *
 * Takes no arguments.
 *
 * Reads WHOLE stdin as [TestParameters] json.
 * Program stdout is redirected to stderr, which is also where errors are printed.
 * Whole [TestReport] json is printed into stdout.
 */
fun main(args: Array<String>) {
    var exitCode: Int

    try {
        val out = System.out
        System.setOut(System.err)

        val testParameters: TestParameters = InputStreamReader(System.`in`, Charsets.UTF_8).readJson()

        val launcher = LauncherFactory.create()
        val reportBuilder = ReportBuildingListener(testParameters.filterStackTraces)
        launcher.registerTestExecutionListeners(reportBuilder)

        val discoveryRequest = LauncherDiscoveryRequestBuilder().apply {
            val selectors = mutableListOf<DiscoverySelector>()
            selectors.addMapped(testParameters.select.uris, ::selectUri)
            selectors.addMapped(testParameters.select.files, ::selectFile)
            selectors.addMapped(testParameters.select.directories, ::selectDirectory)
            selectors.addMapped(testParameters.select.packages, ::selectPackage)
            selectors.addMapped(testParameters.select.classes, ::selectClass)
            selectors.addMapped(testParameters.select.methods, ::selectMethod)
            selectors.addMapped(testParameters.select.resources, ::selectClasspathResource)
            val classpathRoots = HashSet<Path>()
            for (root in testParameters.select.classpathRoots) {
                classpathRoots.add(Paths.get(root))
            }
            selectors.addAll(selectClasspathRoots(classpathRoots))
            selectors(selectors)

            val filters = mutableListOf<Filter<*>>()
            testParameters.filter.engines.included.ifNotEmpty { filters.add(includeEngines(it)) }
            testParameters.filter.engines.excluded.ifNotEmpty { filters.add(excludeEngines(it)) }
            testParameters.filter.classNamePatterns.included.ifNotEmpty { filters.add(includeClassNamePatterns(*it.toTypedArray())) }
            testParameters.filter.classNamePatterns.excluded.ifNotEmpty { filters.add(excludeClassNamePatterns(*it.toTypedArray())) }
            testParameters.filter.packages.included.ifNotEmpty { filters.add(includePackageNames(it)) }
            testParameters.filter.packages.excluded.ifNotEmpty { filters.add(excludePackageNames(it)) }
            testParameters.filter.tags.included.ifNotEmpty { filters.add(includeTags(it)) }
            testParameters.filter.tags.excluded.ifNotEmpty { filters.add(excludeTags(it)) }
            filters(*filters.toTypedArray())
        }.build()

        launcher.execute(discoveryRequest)

        // Output
        if (!reportBuilder.complete) {
            System.err.println("Test report is not complete!")
        }
        val report = reportBuilder.testReport()

        OutputStreamWriter(out, Charsets.UTF_8).use { writer ->
            writer.append(TEST_LAUNCHER_OUTPUT_PREFIX)
            writer.writeJson(report, TestReport::class.java)
            writer.flush()
        }

        exitCode = 0
    } catch (e: Throwable) {
        System.err.println("Exception while running tests")
        e.printStackTrace(System.err)

        exitCode = 1
    } finally {
        System.out.flush()
        System.err.flush()
    }

    System.exit(exitCode)
}

/**
 * JUnit Platform [TestExecutionListener] for collecting test execution data for Wemi.
 */
private class ReportBuildingListener(val filterStackTraces: Boolean) : TestExecutionListener {

    private val testReport = LinkedHashMap<TestIdentifier, TestData>()
    private val startTimes = HashMap<TestIdentifier, Long>()
    var complete = false
        private set

    override fun testPlanExecutionFinished(testPlan: TestPlan?) {
        complete = true
    }

    private fun TestIdentifier.data(): TestData {
        return testReport.getOrPut(this) { TestData() }
    }

    override fun executionSkipped(testIdentifier: TestIdentifier, reason: String?) {
        testIdentifier.data().apply {
            status = TestStatus.SKIPPED
            skipReason = reason
        }
    }

    override fun executionStarted(testIdentifier: TestIdentifier) {
        startTimes[testIdentifier] = System.currentTimeMillis()
    }

    private fun createStackTraceMapper(source: TestSource): ((Array<StackTraceElement>) -> List<StackTraceElement>)? {
        val className: String = when (source) {
            is MethodSource -> source.className ?: return null
            is ClassSource -> source.className ?: return null
            else -> return null
        }
        return { originalStackTrace ->
            // Filters JUnit innards
            var filtered = originalStackTrace.dropLastWhile {
                it.className != className
            }
            // Filter Assertions/Assumptions extras (it throws from deep inside and it is not relevant)
            filtered = filtered.dropWhile { it.className.startsWith("org.junit.jupiter.api.Ass") }


            if (filtered.isEmpty()) {
                originalStackTrace.toList()
            } else {
                filtered
            }
        }
    }

    override fun executionFinished(testIdentifier: TestIdentifier, testExecutionResult: TestExecutionResult) {
        testIdentifier.data().apply {
            duration = startTimes.remove(testIdentifier)?.let { System.currentTimeMillis() - it } ?: -1L
            status = when (testExecutionResult.status) {
                SUCCESSFUL -> TestStatus.SUCCESSFUL
                ABORTED -> TestStatus.ABORTED
                FAILED -> TestStatus.FAILED
                else -> throw IllegalArgumentException("unknown status: ${testExecutionResult.status}")
            }
            val throwable = testExecutionResult.throwable.orElse(null)
            if (throwable != null) {
                val filter = if (filterStackTraces && testIdentifier.source.isPresent)
                    createStackTraceMapper(testIdentifier.source.get())
                else null

                val stackTrace = StringBuilder()
                if (filter == null) {
                    throwable.printStackTrace(StringBuilderWriter(stackTrace))
                } else {
                    stackTrace.appendWithStackTrace(throwable, filter)
                }
                // Drop trailing newline/whitespace
                while (stackTrace.isNotEmpty() && stackTrace.last().isWhitespace()) {
                    stackTrace.setLength(stackTrace.length - 1)
                }
                this.stackTrace = stackTrace.toString()
            }
        }
    }

    override fun reportingEntryPublished(testIdentifier: TestIdentifier, entry: ReportEntry) {
        testIdentifier.data().apply {
            val timestamp = entry.timestamp.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
            reports.ensureCapacity(entry.keyValuePairs.size)
            for ((k, v) in entry.keyValuePairs) {
                reports.add(TestData.ReportEntry(timestamp, k, v))
            }
        }
    }

    fun testReport(): TestReport {
        val result = TestReport()
        testReport.forEach { k, v ->
            result[k.toWemi()] = v
        }
        return result
    }

    private fun TestIdentifier.toWemi(): wemi.test.TestIdentifier {
        return wemi.test.TestIdentifier(
                uniqueId, parentId.orElse(null), displayName,
                isTest, isContainer, tags.asSequence().map { it.name }.toSet(), source.orElse(null)?.toString())
    }
}

private inline fun <T, F> MutableCollection<T>.addMapped(from: Collection<F>, mapper: (F) -> T) {
    for (f in from) {
        this.add(mapper(f))
    }
}

private inline fun <C> C.ifNotEmpty(action: (C) -> Unit) where C : Collection<*> {
    if (this.isNotEmpty()) {
        action(this)
    }
}
